File: julia/specialized/GrobnerSmart.jl
module GrobnerSmart
    import Base: ==, hash
    export Term, Polynomial, TermOrder, set_term_order, set_modulus, grobner_basis, s_polynomial, reduce_polynomial

    mutable struct Term
        coefficient::UInt64
        exponents::UInt64
    end

    ==(t1::Term, t2::Term) = t1.coefficient == t2.coefficient && t1.exponents == t2.exponents
    hash(t::Term, h::UInt) = hash(t.coefficient, hash(t.exponents, h))

    @enum TermOrder Lex GrLex RevLex

    const TERM_ORDER = Ref(Lex)
    const MODULUS = Ref{UInt64}(7)

    function set_term_order(order::TermOrder)
        TERM_ORDER[] = order
    end

    function set_modulus(m::UInt64)
        MODULUS[] = m
    end

    function get_modulus()
        return MODULUS[]
    end

    function mod_inverse(a::UInt64, m::UInt64)
        a_i64 = Int64(a)
        m_i64 = Int64(m)
        m0 = m_i64
        y = 0
        x = 1

        if m == 1
            return UInt64(0)
        end

        while a_i64 > 1
            q = div(a_i64, m_i64)
            t = m_i64

            m_i64 = a_i64 % m_i64
            a_i64 = t
            t = y

            y = x - q * y
            x = t
        end

        if x < 0
            x = x + m0
        end

        return UInt64(x % m0)
    end

    function compare(t1::Term, t2::Term)
        order = TERM_ORDER[]
        if order == Lex
            e1 = t1.exponents & 0x0000FFFFFFFFFFFF
            e2 = t2.exponents & 0x0000FFFFFFFFFFFF
            return cmp(e1, e2)
        elseif order == GrLex
            d1 = (t1.exponents >> 48) & 0xFFFF
            d2 = (t2.exponents >> 48) & 0xFFFF
            if d1 != d2
                return cmp(d1, d2)
            else
                e1 = t1.exponents & 0x0000FFFFFFFFFFFF
                e2 = t2.exponents & 0x0000FFFFFFFFFFFF
                return cmp(e1, e2)
            end
        elseif order == RevLex
            # We don't strictly use RevLex in benchmark natively, but falling back for now
            d1 = (t1.exponents >> 48) & 0xFFFF
            d2 = (t2.exponents >> 48) & 0xFFFF
            if d1 != d2
                return cmp(d1, d2)
            else
                e1 = t1.exponents & 0x0000FFFFFFFFFFFF
                e2 = t2.exponents & 0x0000FFFFFFFFFFFF
                return cmp(e2, e1) # crude revlex approximation
            end
        end
    end

    mutable struct Polynomial
        terms::Vector{Term}
        function Polynomial(terms::Vector{Term})
            sort!(terms, by=t -> t, lt= (a,b) -> compare(a,b) == 1)
            filter!(t -> t.coefficient != 0, terms)
            new(terms)
        end
    end

    ==(p1::Polynomial, p2::Polynomial) = p1.terms == p2.terms
    hash(p::Polynomial, h::UInt) = hash(p.terms, h)

    import Base: +, -, *

    function +(p1::Polynomial, p2::Polynomial)
        result = deepcopy(p1.terms)
        m = MODULUS[]
        for term in p2.terms
            found = false
            for res_term in result
                if res_term.exponents == term.exponents
                    res_term.coefficient = (res_term.coefficient + term.coefficient) % m
                    found = true
                    break
                end
            end
            if !found
                push!(result, deepcopy(term))
            end
        end
        Polynomial(result)
    end

    function -(p1::Polynomial, p2::Polynomial)
        m = MODULUS[]
        neg_p2_terms = [Term((m - (term.coefficient % m)) % m, term.exponents) for term in p2.terms]
        p1 + Polynomial(neg_p2_terms)
    end

    function *(p1::Polynomial, p2::Polynomial)
        new_terms = Term[]
        m = MODULUS[]
        for t1 in p1.terms
            for t2 in p2.terms
                coeff = (t1.coefficient * t2.coefficient) % m
                exps = t1.exponents + t2.exponents
                push!(new_terms, Term(coeff, exps))
            end
        end
        # combine like terms
        combined_terms = Term[]
        for term in new_terms
            found = false
            for c_term in combined_terms
                if c_term.exponents == term.exponents
                    c_term.coefficient = (c_term.coefficient + term.coefficient) % m
                    found = true
                    break
                end
            end
            if !found
                push!(combined_terms, term)
            end
        end
        Polynomial(combined_terms)
    end

    function leading_term(p::Polynomial)
        if isempty(p.terms)
            return Term(UInt64(0), UInt64(0))
        end
        return p.terms[1]
    end

    function get_lcm(a::UInt64, b::UInt64)
        self_exponents = a & 0x0000FFFFFFFFFFFF
        other_exponents = b & 0x0000FFFFFFFFFFFF
        
        lcm_exponents = UInt64(0)
        deg = UInt64(0)
        
        for i in 0:8:40
            self_exp = (self_exponents >> i) & 0xFF
            other_exp = (other_exponents >> i) & 0xFF
            lcm_exp = max(self_exp, other_exp)
            lcm_exponents |= (lcm_exp << i)
            deg += lcm_exp
        end
        
        lcm_exponents |= (deg & 0xFFFF) << 48
        return lcm_exponents
    end

    function can_reduce(a::UInt64, b::UInt64)
        for i in 0:8:40
            self_exp = (a >> i) & 0xFF
            divisor_exp = (b >> i) & 0xFF
            if self_exp < divisor_exp
                return false
            end
        end
        return true
    end

    function s_polynomial(p1::Polynomial, p2::Polynomial)
        lt1 = leading_term(p1)
        lt2 = leading_term(p2)

        if isempty(p1.terms) || isempty(p2.terms)
            return Polynomial(Term[])
        end

        lcm_exponents = get_lcm(lt1.exponents, lt2.exponents)

        s1_exp = lcm_exponents - lt1.exponents
        s2_exp = lcm_exponents - lt2.exponents

        m = MODULUS[]
        inv1 = mod_inverse(lt1.coefficient, m)
        inv2 = mod_inverse(lt2.coefficient, m)

        s1_term = Term(lt2.coefficient, s1_exp)
        s2_term = Term(lt1.coefficient, s2_exp)

        term_poly1 = Polynomial([s1_term])
        term_poly2 = Polynomial([s2_term])

        return (p1 * term_poly1) - (p2 * term_poly2)
    end

    function make_monic(p::Polynomial)
        if isempty(p.terms)
            return p
        end
        m = MODULUS[]
        lt_coeff = leading_term(p).coefficient
        inv_coeff = mod_inverse(lt_coeff, m)
        new_terms = [Term((t.coefficient * inv_coeff) % m, t.exponents) for t in p.terms]
        return Polynomial(new_terms)
    end

    function reduce_polynomial(p::Polynomial, basis::Vector{Polynomial})
        poly = deepcopy(p)
        m = MODULUS[]
        remainder = Term[]

        while !isempty(poly.terms)
            reduced = false
            lt_poly = leading_term(poly)

            for b in basis
                if isempty(b.terms)
                    continue
                end
                lt_b = leading_term(b)
                if can_reduce(lt_poly.exponents, lt_b.exponents)
                    inv_b = mod_inverse(lt_b.coefficient, m)
                    scale_coeff = (lt_poly.coefficient * inv_b) % m
                    scale_exps = lt_poly.exponents - lt_b.exponents

                    scale_term = Term(scale_coeff, scale_exps)
                    scaled_b = b * Polynomial([scale_term])
                    poly = poly - scaled_b
                    reduced = true
                    break
                end
            end

            if !reduced
                push!(remainder, deepcopy(leading_term(poly)))
                popfirst!(poly.terms)
            end
        end
        return Polynomial(remainder)
    end

    function grobner_basis(polynomials::Vector{Polynomial})
        basis = deepcopy(polynomials)
        basis_set = Set{Polynomial}(basis)
        n_orig = length(basis)
        pair_set = [(i, j) for i in 1:length(basis) for j in (i+1):length(basis)]

        while !isempty(pair_set)
            (i, j) = popfirst!(pair_set)
            s_poly = s_polynomial(basis[i], basis[j])
            h = reduce_polynomial(s_poly, basis)

            if !isempty(h.terms) && !in(h, basis_set)
                push!(basis, h)
                push!(basis_set, h)
                n = length(basis)
                for k in 1:(n-1)
                    push!(pair_set, (k, n))
                end
            end
        end

        reduced_basis = Polynomial[]
        for i in 1:length(basis)
            basis_excluding_self = vcat(basis[1:i-1], basis[i+1:end])
            reduced = reduce_polynomial(basis[i], basis_excluding_self)
            if !isempty(reduced.terms)
                push!(reduced_basis, make_monic(reduced))
            end
        end

        return reduced_basis
    end

    function pack_exponents(exponents::Vector{Int})
        packed = UInt64(0)
        padded_exps = copy(exponents)
        while length(padded_exps) < 6
            push!(padded_exps, 0)
        end
        for i in 1:6
            exp = padded_exps[i]
            shift = UInt64(40 - 8 * (i - 1))
            packed |= (UInt64(exp) & 0xFF) << shift
        end
        deg = UInt64(sum(padded_exps[1:6]))
        packed |= (deg & 0xFFFF) << 48
        return packed
    end

    function main(n::Int)
        set_modulus(UInt64(7))
        polynomials = Polynomial[]
        if n == 4
            println("Julia specialized bit-packed cyclic 4")
            p1 = Polynomial([
                Term(UInt64(1), pack_exponents([1, 0, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 1, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 1, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 0, 1])),
            ])
            p2 = Polynomial([
                Term(UInt64(1), pack_exponents([1, 1, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 1, 1, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 0, 0, 1])),
            ])
            p3 = Polynomial([
                Term(UInt64(1), pack_exponents([1, 1, 1, 0])),
                Term(UInt64(1), pack_exponents([0, 1, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 0, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 1, 0, 1])),
            ])
            p4 = Polynomial([
                Term(UInt64(1), pack_exponents([1, 1, 1, 1])),
                Term(UInt64(get_modulus() - 1), pack_exponents([0, 0, 0, 0])),
            ])
            polynomials = [p1, p2, p3, p4]
        end
        if n == 5
            println("Julia specialized bit-packed cyclic 5")
            p1 = Polynomial([
                Term(UInt64(1), pack_exponents([1, 0, 0, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 1, 0, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 1, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 0, 1, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 0, 0, 1])),
            ])
            p2 = Polynomial([
                Term(UInt64(1), pack_exponents([1, 1, 0, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 1, 1, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 1, 1, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 0, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 0, 0, 0, 1])),
            ])
            p3 = Polynomial([
                Term(UInt64(1), pack_exponents([1, 1, 1, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 1, 1, 1, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 1, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 0, 0, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 1, 0, 0, 1])),
            ])
            p4 = Polynomial([
                Term(UInt64(1), pack_exponents([1, 1, 1, 1, 0])),
                Term(UInt64(1), pack_exponents([0, 1, 1, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 0, 1, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 1, 0, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 1, 1, 0, 1])),
            ])
            p5 = Polynomial([
                Term(UInt64(1), pack_exponents([1, 1, 1, 1, 1])),
                Term(UInt64(get_modulus() - 1), pack_exponents([0, 0, 0, 0, 0])),
            ])

            polynomials = [p1, p2, p3, p4, p5]
        end
        if n == 6
            println("Julia specialized bit-packed cyclic 6")
            p1 = Polynomial([
                Term(UInt64(1), pack_exponents([1, 0, 0, 0, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 1, 0, 0, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 1, 0, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 0, 1, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 0, 0, 1, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 0, 0, 0, 1])),
            ])
            p2 = Polynomial([
                Term(UInt64(1), pack_exponents([1, 1, 0, 0, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 1, 1, 0, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 1, 1, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 0, 1, 1, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 0, 0, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 0, 0, 0, 0, 1])),
            ])
            p3 = Polynomial([
                Term(UInt64(1), pack_exponents([1, 1, 1, 0, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 1, 1, 1, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 1, 1, 1, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 0, 1, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 0, 0, 0, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 1, 0, 0, 0, 1])),
            ])
            p4 = Polynomial([
                Term(UInt64(1), pack_exponents([1, 1, 1, 1, 0, 0])),
                Term(UInt64(1), pack_exponents([0, 1, 1, 1, 1, 0])),
                Term(UInt64(1), pack_exponents([0, 0, 1, 1, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 0, 0, 1, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 1, 0, 0, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 1, 1, 0, 0, 1])),
            ])
            p5 = Polynomial([
                Term(UInt64(1), pack_exponents([1, 1, 1, 1, 1, 0])),
                Term(UInt64(1), pack_exponents([0, 1, 1, 1, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 0, 1, 1, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 1, 0, 1, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 1, 1, 0, 1, 1])),
                Term(UInt64(1), pack_exponents([1, 1, 1, 1, 0, 1])),
            ])
            p6 = Polynomial([
                Term(UInt64(1), pack_exponents([1, 1, 1, 1, 1, 1])),
                Term(UInt64(get_modulus() - 1), pack_exponents([0, 0, 0, 0, 0, 0])),
            ])

            polynomials = [p1, p2, p3, p4, p5, p6]
        end

        set_term_order(Lex)
        for i in 0:9 
            basis = grobner_basis(polynomials)
            println("Iteration $i complete")
            #if i == 9
                #println("Grobner basis computed with size: ", length(basis))
            #end
        end
        
    end
end

if abspath(PROGRAM_FILE) == @__FILE__
    n = 4 # Default to cyclic 4
    if length(ARGS) > 0
        n = parse(Int, ARGS[1])
    end
    GrobnerSmart.main(n)
end

File: cpp/specialized/GrobnerSmart.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <chrono>
#include <unordered_set>

using namespace std;

enum TermOrder { Lex, GrLex, RevLex };
TermOrder TERM_ORDER = Lex;
uint64_t MODULUS = 7;

uint64_t mod_inverse(uint64_t a, uint64_t m) {
    if (m == 1) return 0;
    int64_t a_i64 = a;
    int64_t m_i64 = m;
    int64_t m0 = m_i64;
    int64_t y = 0, x = 1;
    while (a_i64 > 1) {
        int64_t q = a_i64 / m_i64;
        int64_t t = m_i64;
        m_i64 = a_i64 % m_i64;
        a_i64 = t;
        t = y;
        y = x - q * y;
        x = t;
    }
    if (x < 0) x += m0;
    return (uint64_t)(x % m0);
}

struct Term {
    uint64_t coefficient;
    uint64_t exponents;
};

int compare_terms(const Term& t1, const Term& t2) {
    uint64_t e1 = t1.exponents & 0x0000FFFFFFFFFFFFULL;
    uint64_t e2 = t2.exponents & 0x0000FFFFFFFFFFFFULL;
    if (e1 < e2) return -1;
    if (e1 > e2) return 1;
    return 0;
}

class Polynomial {
public:
    vector<Term> terms;
    Polynomial() {}
    Polynomial(vector<Term> _terms) {
        for (auto& t : _terms) {
            if (t.coefficient != 0) {
                terms.push_back(t);
            }
        }
        sort(terms.begin(), terms.end(), [](const Term& a, const Term& b) {
            return compare_terms(a, b) > 0;
        });
    }
};

Polynomial add_poly(const Polynomial& p1, const Polynomial& p2) {
    vector<Term> result = p1.terms;
    for (const auto& term2 : p2.terms) {
        bool found = false;
        for (auto& res_term : result) {
            if (res_term.exponents == term2.exponents) {
                res_term.coefficient = (res_term.coefficient + term2.coefficient) % MODULUS;
                found = true;
                break;
            }
        }
        if (!found) result.push_back(term2);
    }
    return Polynomial(result);
}

Polynomial sub_poly(const Polynomial& p1, const Polynomial& p2) {
    vector<Term> result = p1.terms;
    for (const auto& term : p2.terms) {
        bool found = false;
        for (auto& res_term : result) {
            if (res_term.exponents == term.exponents) {
                res_term.coefficient = (MODULUS + res_term.coefficient - term.coefficient) % MODULUS;
                found = true;
                break;
            }
        }
        if (!found) {
            result.push_back({(MODULUS - (term.coefficient % MODULUS)) % MODULUS, term.exponents});
        }
    }
    return Polynomial(result);
}


Polynomial multiply_by_term(const Polynomial& p, const Term& term) {
    vector<Term> new_terms;
    new_terms.reserve(p.terms.size());
    for (const auto& t : p.terms) {
        new_terms.push_back({(t.coefficient * term.coefficient) % MODULUS, t.exponents + term.exponents});
    }
    return Polynomial(new_terms);
}


Term leading_term(const Polynomial& p) {
    if (p.terms.empty()) return {0, 0};
    return p.terms[0];
}

uint64_t get_lcm(uint64_t a, uint64_t b) {
    uint64_t self_exponents = a & 0x0000FFFFFFFFFFFFULL;
    uint64_t other_exponents = b & 0x0000FFFFFFFFFFFFULL;
    uint64_t lcm_exponents = 0;
    uint64_t deg = 0;
    for (int i = 0; i <= 40; i += 8) {
        uint64_t self_exp = (self_exponents >> i) & 0xFF;
        uint64_t other_exp = (other_exponents >> i) & 0xFF;
        uint64_t lcm_exp = max(self_exp, other_exp);
        lcm_exponents |= (lcm_exp << i);
        deg += lcm_exp;
    }
    lcm_exponents |= (deg & 0xFFFF) << 48;
    return lcm_exponents;
}

bool can_reduce(uint64_t a, uint64_t b) {
    for (int i = 0; i <= 40; i += 8) {
        uint64_t self_exp = (a >> i) & 0xFF;
        uint64_t divisor_exp = (b >> i) & 0xFF;
        if (self_exp < divisor_exp) return false;
    }
    return true;
}

Polynomial s_polynomial(const Polynomial& p1, const Polynomial& p2) {
    Term lt1 = leading_term(p1);
    Term lt2 = leading_term(p2);
    if (p1.terms.empty() || p2.terms.empty()) return Polynomial();

    uint64_t lcm_exps = get_lcm(lt1.exponents, lt2.exponents);
    uint64_t s1_exp = lcm_exps - lt1.exponents;
    uint64_t s2_exp = lcm_exps - lt2.exponents;

    Term s1_term = {lt2.coefficient, s1_exp};
    Term s2_term = {lt1.coefficient, s2_exp};

    return sub_poly(multiply_by_term(p1, s1_term), multiply_by_term(p2, s2_term));
}

Polynomial make_monic(const Polynomial& p) {
    if (p.terms.empty()) return p;
    uint64_t lt_coeff = leading_term(p).coefficient;
    uint64_t inv_coeff = mod_inverse(lt_coeff, MODULUS);
    vector<Term> new_terms;
    for (const auto& t : p.terms) {
        new_terms.push_back({(t.coefficient * inv_coeff) % MODULUS, t.exponents});
    }
    return Polynomial(new_terms);
}

Polynomial reduce_polynomial(const Polynomial& p, const vector<Polynomial>& basis) {
    Polynomial poly = p;
    vector<Term> remainder;
    while (!poly.terms.empty()) {
        bool reduced = false;
        Term lt_poly = leading_term(poly);
        for (const auto& b : basis) {
            if (b.terms.empty()) continue;
            Term lt_b = leading_term(b);
            if (can_reduce(lt_poly.exponents, lt_b.exponents)) {
                uint64_t inv_b = mod_inverse(lt_b.coefficient, MODULUS);
                uint64_t scale_coeff = (lt_poly.coefficient * inv_b) % MODULUS;
                uint64_t scale_exps = lt_poly.exponents - lt_b.exponents;
                
                Polynomial scaled_b = multiply_by_term(b, {scale_coeff, scale_exps});
                poly = sub_poly(poly, scaled_b);
                reduced = true;
                break;
            }
        }
        if (!reduced) {
            remainder.push_back(leading_term(poly));
            poly.terms.erase(poly.terms.begin());
        }
    }
    return Polynomial(remainder);
}


bool polynomials_equal(const Polynomial& p1, const Polynomial& p2) {
    if (p1.terms.size() != p2.terms.size()) return false;
    for (size_t i = 0; i < p1.terms.size(); i++) {
        if (p1.terms[i].coefficient != p2.terms[i].coefficient || p1.terms[i].exponents != p2.terms[i].exponents) {
            return false;
        }
    }
    return true;
}

inline bool operator==(const Polynomial& p1, const Polynomial& p2) {
    return polynomials_equal(p1, p2);
}

struct PolyHash {
    size_t operator()(const Polynomial& p) const {
        size_t hash = 0;
        for (const auto& t : p.terms) {
            hash ^= t.coefficient + 0x9e3779b9 + (hash << 6) + (hash >> 2);
            hash ^= t.exponents + 0x9e3779b9 + (hash << 6) + (hash >> 2);
        }
        return hash;
    }
};

vector<Polynomial> grobner_basis(vector<Polynomial> basis) {
    unordered_set<Polynomial, PolyHash> basis_set;
    for (const auto& poly : basis) {
        basis_set.insert(poly);
    }

    vector<pair<int, int>> pair_set;
    for (size_t i = 0; i < basis.size(); i++) {
        for (size_t j = i + 1; j < basis.size(); j++) {
            pair_set.push_back({(int)i, (int)j});
        }
    }
    while (!pair_set.empty()) {
        auto p = pair_set.front();
        pair_set.erase(pair_set.begin());
        
        Polynomial s_poly = s_polynomial(basis[p.first], basis[p.second]);
        Polynomial h = reduce_polynomial(s_poly, basis);
        
        if (!h.terms.empty() && basis_set.insert(h).second) {
            int n = basis.size();
            basis.push_back(h);
            for (int k = 0; k < n; k++) {
                pair_set.push_back({k, n});
            }
        }
    }
    vector<Polynomial> reduced_basis;
    for (size_t i = 0; i < basis.size(); i++) {
        vector<Polynomial> basis_excluding_self;
        for (size_t j = 0; j < basis.size(); j++) {
            if (i != j) basis_excluding_self.push_back(basis[j]);
        }
        Polynomial reduced = reduce_polynomial(basis[i], basis_excluding_self);
        if (!reduced.terms.empty()) {
            reduced_basis.push_back(make_monic(reduced));
        }
    }
    return reduced_basis;
}

uint64_t pack_exponents(vector<int> exps) {
    uint64_t packed = 0;
    while(exps.size() < 6) exps.push_back(0);
    for (int i = 0; i < 6; i++) {
        uint64_t shift = 40 - 8 * i;
        packed |= ((uint64_t)exps[i] & 0xFF) << shift;
    }
    uint64_t deg = 0;
    for (int i = 0; i < 6; i++) deg += exps[i];
    packed |= (deg & 0xFFFF) << 48;
    return packed;
}

int main(int argc, char** argv) {
    int n = 4;
    if (argc > 1) {
        n = stoi(argv[1]);
    }
    
    MODULUS = 7;
    TERM_ORDER = Lex;
    vector<Polynomial> polynomials;

    if (n == 4) {
        cout << "C++ specialized bit-packed cyclic 4\n";
        polynomials = {
            Polynomial({ {1, pack_exponents({1, 0, 0, 0})}, {1, pack_exponents({0, 1, 0, 0})}, {1, pack_exponents({0, 0, 1, 0})}, {1, pack_exponents({0, 0, 0, 1})} }),
            Polynomial({ {1, pack_exponents({1, 1, 0, 0})}, {1, pack_exponents({0, 1, 1, 0})}, {1, pack_exponents({0, 0, 1, 1})}, {1, pack_exponents({1, 0, 0, 1})} }),
            Polynomial({ {1, pack_exponents({1, 1, 1, 0})}, {1, pack_exponents({0, 1, 1, 1})}, {1, pack_exponents({1, 0, 1, 1})}, {1, pack_exponents({1, 1, 0, 1})} }),
            Polynomial({ {1, pack_exponents({1, 1, 1, 1})}, {6, pack_exponents({0, 0, 0, 0})} })
        };
    } else if (n == 5) {
        cout << "C++ specialized bit-packed cyclic 5\n";
        polynomials = {
            Polynomial({ {1, pack_exponents({1, 0, 0, 0, 0})}, {1, pack_exponents({0, 1, 0, 0, 0})}, {1, pack_exponents({0, 0, 1, 0, 0})}, {1, pack_exponents({0, 0, 0, 1, 0})}, {1, pack_exponents({0, 0, 0, 0, 1})} }),
            Polynomial({ {1, pack_exponents({1, 1, 0, 0, 0})}, {1, pack_exponents({0, 1, 1, 0, 0})}, {1, pack_exponents({0, 0, 1, 1, 0})}, {1, pack_exponents({0, 0, 0, 1, 1})}, {1, pack_exponents({1, 0, 0, 0, 1})} }),
            Polynomial({ {1, pack_exponents({1, 1, 1, 0, 0})}, {1, pack_exponents({0, 1, 1, 1, 0})}, {1, pack_exponents({0, 0, 1, 1, 1})}, {1, pack_exponents({1, 0, 0, 1, 1})}, {1, pack_exponents({1, 1, 0, 0, 1})} }),
            Polynomial({ {1, pack_exponents({1, 1, 1, 1, 0})}, {1, pack_exponents({0, 1, 1, 1, 1})}, {1, pack_exponents({1, 0, 1, 1, 1})}, {1, pack_exponents({1, 1, 0, 1, 1})}, {1, pack_exponents({1, 1, 1, 0, 1})} }),
            Polynomial({ {1, pack_exponents({1, 1, 1, 1, 1})}, {6, pack_exponents({0, 0, 0, 0, 0})} })
        };
    } else if (n == 6) {
        cout << "C++ specialized bit-packed cyclic 6\n";
        polynomials = {
            Polynomial({ {1, pack_exponents({1, 0, 0, 0, 0, 0})}, {1, pack_exponents({0, 1, 0, 0, 0, 0})}, {1, pack_exponents({0, 0, 1, 0, 0, 0})}, {1, pack_exponents({0, 0, 0, 1, 0, 0})}, {1, pack_exponents({0, 0, 0, 0, 1, 0})}, {1, pack_exponents({0, 0, 0, 0, 0, 1})} }),
            Polynomial({ {1, pack_exponents({1, 1, 0, 0, 0, 0})}, {1, pack_exponents({0, 1, 1, 0, 0, 0})}, {1, pack_exponents({0, 0, 1, 1, 0, 0})}, {1, pack_exponents({0, 0, 0, 1, 1, 0})}, {1, pack_exponents({0, 0, 0, 0, 1, 1})}, {1, pack_exponents({1, 0, 0, 0, 0, 1})} }),
            Polynomial({ {1, pack_exponents({1, 1, 1, 0, 0, 0})}, {1, pack_exponents({0, 1, 1, 1, 0, 0})}, {1, pack_exponents({0, 0, 1, 1, 1, 0})}, {1, pack_exponents({0, 0, 0, 1, 1, 1})}, {1, pack_exponents({1, 0, 0, 0, 1, 1})}, {1, pack_exponents({1, 1, 0, 0, 0, 1})} }),
            Polynomial({ {1, pack_exponents({1, 1, 1, 1, 0, 0})}, {1, pack_exponents({0, 1, 1, 1, 1, 0})}, {1, pack_exponents({0, 0, 1, 1, 1, 1})}, {1, pack_exponents({1, 0, 0, 1, 1, 1})}, {1, pack_exponents({1, 1, 0, 0, 1, 1})}, {1, pack_exponents({1, 1, 1, 0, 0, 1})} }),
            Polynomial({ {1, pack_exponents({1, 1, 1, 1, 1, 0})}, {1, pack_exponents({0, 1, 1, 1, 1, 1})}, {1, pack_exponents({1, 0, 1, 1, 1, 1})}, {1, pack_exponents({1, 1, 0, 1, 1, 1})}, {1, pack_exponents({1, 1, 1, 0, 1, 1})}, {1, pack_exponents({1, 1, 1, 1, 0, 1})} }),
            Polynomial({ {1, pack_exponents({1, 1, 1, 1, 1, 1})}, {6, pack_exponents({0, 0, 0, 0, 0, 0})} })
        };
    }

    for (int i = 0; i < 10; ++i) {
        auto basis = grobner_basis(polynomials);
        cout << "Iteration " << i << " complete\n";
    }

    return 0;
}
